### PR TITLE
Timeout warning

### PR DESCRIPTION
**Pull request by sequencemedialimited**
[Trello card](https://trello.com/c/MA4PWU4W/206-show-user-warning-of-session-timeout)

## Context

Some of the forms have quite a lot of expected input and users tend to use the browser to write a lot of content expecting it to save their progress.

Users should be told when they may be about to lose content that they have been putting directly into the form.

We should show users a warning when they are getting close to their session timing out

### Javascript solution

When the user loads a page, the server records this as the last interaction in the session.

5 minutes before the time finishes, the modal is open. If user dismisses the modal, client notifies server of new time user was active and server extends session appropriately.

If user does not interact with the modal for another 5 minutes, client will redirect user appropriately to the timeout page.

## No javascript solution

Use a meta refresh tag inside Noscript tags that matches the session timeout time. If user does not submit the form by the specified time, meta refresh will take them to the timeout page.

### Caveats

Should not limit the number of times users can extend their session.
The save and return is not working so we will only add the continue application button to the modal.

### CODE REVIEW

File: metadata/config/service.json [hunk 1]
```diff
@@ -8,5 +8,5 @@
   "phaseText": "Service phase text goes here",
   "serviceEmailAddress": "form-builder@digital.justice.gov.uk",
   "serviceUrl": "page.start",
-  "sessionDuration": 30
-}
\ No newline at end of file
+  "sessionDuration": 1
```
Comment (brenetic) [inline metadata/config/service.json]: Should we swap this back to 30?

File: specifications/timeout-warning/templates/nunjucks/template.njk [hunk 2]
```diff
@@ -0,0 +1,28 @@
+{%- macro timeoutWarning(sessionDuration) %}
+
+<div class="govuk-timeout-warning-fallback">
+  <p>We will reset your application if you do not complete the form and press Next
+  by {{ sessionDuration }} minutes.</p>
+  <p>We do this to keep your information secure.</p>
```
Comment (brenetic) [inline specifications/timeout-warning/templates/nunjucks/template.njk]: Have we run the content past Cogsy yet? I guess also the content of the Session Timed out page

Comment (tomas-stefano) [reply specifications/timeout-warning/templates/nunjucks/template.njk]: I will pass into him about the content probably tomorrow. I will let you know when it is approved.

File: src/js/component/timeout-warning.js [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+
+TimeoutWarning.prototype.escClose = function () {
+ if (this.isDialogOpen() && event.keyCode === 27) {
+   this.closeDialog()
+ }
+}
+
+// Copied from gov.uk accessible timeout
+TimeoutWarning.prototype.numberToWords = function () {
+  var string = n.toString()
```
Comment (sequencemedialimited) [inline src/js/component/timeout-warning.js]: `n` is not defined ;)

**Comment by sequencemedialimited** [comment]
I've merged with the latest master. This should only modify _latest built assets_ and not touch your changes, @tomas-stefano

**Comment by tomas-stefano** [comment]
Testing the comment notification on slack now 🥇

File: specifications/timeout-warning/templates/nunjucks/template.njk (hunk 2 above)
Comment (tomas-stefano) [reply specifications/timeout-warning/templates/nunjucks/template.njk]: Now the content is being renewed (Phill Cogger sent the final content) :)

**Pull request merged**